README.md
# EncFS
EncFS binaries for the Android environment, packaged as Magisk module

Built by [DualJoe](https://forum.xda-developers.com/member.php?u=4293606) @ XDA Developers

## Description
Installing this module will allow you to have EncFS binary and other required binaries in `/system/bin`, thus allowing you to make use of the encryption for your files.

This module only packages the binary files! No feature is provided with this module.

Binary files are taken from "[Linux binaries for Android](https://forum.xda-developers.com/android/software/utils-data-recovery-tools-testdisk-t3709380)" XDA thread by DualJoe.

Note: Only report issues/questions on this GitHub repository and not XDA thread linked above. This module is packed and managed by TechnoSparks and not DualJoe!

## Requirements
- Magisk version 23
- ARM device (32-bit (ARMv7) or 64-bit (ARMv8))

## Notes
- Make sure your "Mount Namespace Mode" is set to "Global", otherwise mounted `encfs` directories won't reflect to other apps!
- Using FUSE on Android require root. Since `encfs` uses FUSE, it inherently needs root as well. Execute encfs in a root shell always.
- Much more useful info can be gathered in [this post](https://forum.xda-developers.com/t/mod-encrypt-sdcard-with-encfs.3027711/#post-75721227)

## About
- This is simply a module that puts the EncFS binaries into Android utilizing [Magisk](https://github.com/topjohnwu/Magisk)
- Binaries are cross-compiled from source to Android by DualJoe @ XDA
- [EncFS source on GitHub](https://github.com/vgough/encfs)

changelog.md
# 1.0.4
- This release prepares the module for the new update mechanism of Magisk v24
- Fixes the one bug that shouldn't even happen in 1.0.2
- Added changelog.md so Magisk Manager can properly show it

customize.sh
SKIPUNZIP=1
# must handle extraction manually

ui_print "- Identifying arch: $ARCH"
if [ "$ARCH" != "arm64" ] && [ "$ARCH" != "arm" ]; then
    ui_print "[!] Unrecognised architecture!"
    ui_print "    Binaries are for ARM only"
    ui_print "    Installation aborted"
    abort
fi

ui_print "- Extracting module files"
unzip -o "$ZIPFILE" -x 'META-INF/*' -d $MODPATH >&2

# Default permissions
set_perm_recursive $MODPATH 0 0 0755 0644
set_perm_recursive $MODPATH/arm 0 2000 0755 0755
set_perm_recursive $MODPATH/arm64 0 2000 0755 0755
#set_perm_recursive $MODPATH/system/xbin 0 2000 0755 0755
#set_perm_recursive $MODPATH/system/system_ext/bin 0 2000 0755 0755
#set_perm_recursive $MODPATH/system/vendor/bin 0 2000 0755 0755 u:object_r:vendor_file:s0

ui_print "- Trimming non-compatible arch"
if [ "$ARCH" == "arm64" ]; then
    rm -rf $MODPATH/arm
    mv $MODPATH/arm64 $MODPATH/system
elif [ "$ARCH" == "arm" ]; then
    rm -rf $MODPATH/arm64
    mv $MODPATH/arm $MODPATH/system
fi

ui_print "- Preserve README!"
mv $MODPATH/README.md $MODPATH/Module-Readme.md

config.sh
##########################################################################################
#
# Magisk Module Template Config Script
# by topjohnwu
#
##########################################################################################
##########################################################################################
#
# Instructions:
#
# 1. Place your files into system folder (delete the placeholder file)
# 2. Fill in your module's info into module.prop
# 3. Configure the settings in this file (config.sh)
# 4. If you need boot scripts, add them into common/post-fs-data.sh or common/service.sh
# 5. Add your additional or modified system properties into common/system.prop
#
##########################################################################################

##########################################################################################
# Configs
##########################################################################################

# Set to true if you need to enable Magic Mount
# Most mods would like it to be enabled
AUTOMOUNT=true

# Set to true if you need to load system.prop
PROPFILE=false

# Set to true if you need post-fs-data script
POSTFSDATA=false

# Set to true if you need late_start service script
LATESTARTSERVICE=false

##########################################################################################
# Installation Message
##########################################################################################

# Set what you want to show when installing your mod

print_modname() {
    ui_print "================================================"
    ui_print " EncFS v1.9.5"
    ui_print "================================================"
    ui_print "Built and provided by DualJoe"
    ui_print "Module proudly made by TechnoSparks"
    ui_print "Powered by Magisk (@topjohnwu)"
    ui_print " "
}

##########################################################################################
# Replace list
##########################################################################################

# List all directories you want to directly replace in the system
# Check the documentations for more info about how Magic Mount works, and why you need this

# This is an example
REPLACE="
/system/app/Youtube
/system/priv-app/SystemUI
/system/priv-app/Settings
/system/framework
"

# Construct your own list here, it will override the example above
# !DO NOT! remove this if you don't need to replace anything, leave it empty as it is now
REPLACE="
"

##########################################################################################
# Permissions
##########################################################################################

set_permissions() {
  # Only some special files require specific permissions
  # The default permissions should be good enough for most cases

  # Here are some examples for the set_perm functions:

  # set_perm_recursive  <dirname>                <owner> <group> <dirpermission> <filepermission> <contexts> (default: u:object_r:system_file:s0)
  # set_perm_recursive  $MODPATH/system/lib       0       0       0755            0644

  # set_perm  <filename>                         <owner> <group> <permission> <contexts> (default: u:object_r:system_file:s0)
  # set_perm  $MODPATH/system/bin/app_process32   0       2000    0755         u:object_r:zygote_exec:s0
  # set_perm  $MODPATH/system/bin/dex2oat         0       2000    0755         u:object_r:dex2oat_exec:s0
  # set_perm  $MODPATH/system/lib/libart.so       0       0       0644

  # The following is default permissions, DO NOT remove
  set_perm_recursive  $MODPATH                0  0  0755  0644
  set_perm_recursive  $MODPATH/system/xbin    0  0  0755  0755
}

##########################################################################################
# Custom Functions
##########################################################################################

# This file (config.sh) will be sourced by the main flash script after util_functions.sh
# If you need custom logic, please add them here as functions, and call these functions in
# update-binary. Refrain from adding code directly into update-binary, as it will make it
# difficult for you to migrate your modules to newer template versions.
# Make update-binary as clean as possible, try to only do function calls in it.

extractAll() {
    if [ $ARCH = "arm64" ]; then
        ui_print "- Extracting arm64 binaries"
        unzip -o "$ZIP" 'arm64/*' -d $MODPATH >&2
        mv $MODPATH/arm64 $MODPATH/system
    elif [ $ARCH = "arm" ]; then
        ui_print "- Extracting arm binaries"
        unzip -o "$ZIP" 'arm/*' -d $MODPATH >&2
        mv $MODPATH/arm $MODPATH/system
    else
        ui_print "[!] Unrecognised architecture!"
        ui_print "    Binaries are for ARM only"
        ui_print "------------------------------------------------"
        abort
    fi
    unzip -o "$ZIP" README.md -d $MODPATH >&2
}

print_onFinish() {
    ui_print " "
    ui_print "[i] Flashing is done!"
    ui_print "------------------------------------------------"
}